### PR TITLE
Names! In! Space! Part 1

### PR DESCRIPTION
**Pull request by adamemerson**
Fix the namespace situation.

Like Gaul, this can be divided into four parts:

1. Don't use encode and decode as qualified symbols. Ensure that ceph::encode and ceph::decode are in scope then call them unqualified. Argument Dependent Lookup will take care of the rest.

2. Wrap everything in a namespace. We have some things in and some things out willy-nilly. Core code will go in namespace ceph. Things that use the librados interface will go in their own namespace (rgw, etc.) This way we can pull things in where needed.

3. Include whatever we want to use unqualified in our namespaces. Whether for convenience or to pick one implementation over another or what have you. We can put some subset of this in an inline namespace so that things like RGW can pull them into their environments, too, without getting all of namespace ceph.

4. Remove the using namespace ceph and using namespace std statements from all header files. .cc files it's not actually so important, they can be cleaned up as needed. But having them in header files is starting to cause symbol clashes and it needs to stop.

### CODE REVIEW

File: src/include/encoding.h [hunk 1]
```diff
@@ -14,6 +14,22 @@
...
+#include <set>
+#include <map>
+#include <deque>
+#include <vector>
+#include <string>
+#include <boost/optional/optional_io.hpp>
+#include <boost/tuple/tuple.hpp>
+
+#ifndef _BACKWARD_BACKWARD_WARNING_H
+#define _BACKWARD_BACKWARD_WARNING_H   // make gcc 4.3 shut up about hash_*
```
Comment (tchaikov) [inline src/include/encoding.h]: before switching to gcc-7, we were using gcc 4.8 and up. guess it's a legacy of the C++98 era. can we take this opportunity to drop this `#define`? or make it a separated commit in this PR?

File: src/include/denc.h [hunk 2]
```diff
@@ -1651,8 +1648,8 @@ inline typename std::enable_if<traits::supported &&
 
 // nohead variants
 template<typename T, typename traits=denc_traits<T>>
-inline typename std::enable_if<traits::supported &&
-			       !traits::featured>::type encode_nohead(
+inline std::enable_if_t<traits::supported &&
```
Comment (chardan) [inline src/include/denc.h]: Yay, C++14!

File: src/mds/LogEvent.h [hunk 3]
```diff
@@ -93,7 +93,7 @@ class LogEvent {
     ::encode(EVENT_NEW_ENCODING, bl);
     ENCODE_START(1, 1, bl)
     ::encode(_type, bl);
-    encode(bl, features);
+    this->encode(bl, features);
```
Comment (chardan) [inline src/mds/LogEvent.h]: Agreed, much more clear.

**Comment by adamemerson** [comment]
@chardan The fourth part of Gaul was recently revealed by decrypting ancient messages: https://urbigenous.net/library/alicebob.html

THOUGH! The question I have now is. Is it better to commit things piece by piece? (Say, since this part does 'common' should we just chunk it in now and the like with all the other directories) Or should we do the whole thing as one gigantic enormous PR?

**Comment by chardan** [comment]
@adamemerson I will defer if someone has a strong opinion, but here I think it makes sense to do it all at once, it isn't mostly concerned with runtime.

**Comment by tchaikov** [comment]
@chardan @adamemerson i'd suggest do this piece by piece. this helps with interested reviewer to identify the commit he/she wants to look at.

**Comment by adamemerson** [comment]
Jenkins, retest this please.

**Comment by adamemerson** [comment]
@tchaikov I think you're right. And trying to do it all in one HUGE PR will make conflicts pretty unavoidable.

**Comment by chardan** [comment]
+1 multiple commits, confused myself into thinking "multiple PRs"!

**Comment by tchaikov** [comment]
@adamemerson need rebase.

**Comment by adamemerson** [comment]
@tchaikov All right, rebased!

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-01-11_15:01:24-rados-wip-sage3-testing-2018-01-11-0711-distro-basic-smithi/#

**Comment by liewegas** [comment]
merged manually!

**Pull request closed without merging**